### ISSUE TITLE
Insufficient memory error

### ISSUE THREAD

**Issue by tschennie** [barnhill/barcodelib#50]
I'm using the bardcodelib to generate barcodes within PDFs by RDLC-Designer.
If many people use our system at the same time, sometimes the PDF will be generated, but the barcode on it is missing and we got following error:

```
The definition of the report '' is invalid.
InnerException: 
An unexpected error occurred in Report Processing.
InnerException: 
Insufficient memory to continue the execution of the program.
```

Code snippet:
```
        public byte[] GetBarCode39(string CodeNumber, int Lenght = 1000, int Height = 200, int FontSize = 40)
        {
            byte[] image = null;

            try
            {
                MemoryStream ms = new MemoryStream();
                Barcode barcode = new Barcode();
                barcode.IncludeLabel = true;
                barcode.Alignment = AlignmentPositions.CENTER;
                barcode.LabelFont = new Font(FontFamily.GenericMonospace, FontSize, FontStyle.Regular);

                var barcodeImage = barcode.Encode(TYPE.CODE39, CodeNumber, System.Drawing.Color.Black, System.Drawing.Color.White, Lenght, Height);

                barcodeImage.Save(ms, ImageFormat.Jpeg);

                image = ms.GetBuffer();
            }
            catch (Exception)
            {
                return null;
            }
            return image;
        }
```

We already gave more power to the azure webservices, but this didn't solve the problem.. Do you know what's the problem?

**Comment by binki** [comment]
Have you tried disposing the returned image? I think since [`Bitmap`](https://docs.microsoft.com/en-us/dotnet/api/system.drawing.bitmap?view=netframework-4.7.2) wraps system GDI+ objects, creating a lot without explicitly cleaning them up might exacerbate the situation.

Sorry, I haven’t tried to reproduce the issue, so this is just a guess.

**Comment by binki** [comment]
You could use code like:

```cs
using (var barcodeImage = barcode.Encode(TYPE.CODE39, CodeNumber, System.Drawing.Color.Black, System.Drawing.Color.White, Lenght, Height)) {
  barcodeImage.Save(ms, ImageFormat.Jpeg);
  image = ms.GetBuffer();
}
```

**Comment by barnhill** [comment]
@tschennie I think @binki is headed in the right direction.  Bitmaps will eat up tons of memory if you dont release them.  Ill leave this open till you comment back and let us know if that helps.

**Comment by tschennie** [comment]
Thanks! I will try the code snippet from @binki and come back, if it's solved.

**Comment by barnhill** [comment]
@tschennie keep me posted so I can close this issue out if its solved.

**Comment by tschennie** [comment]
I deployed the code snippet but now we got following error:
`ELABEL_GENERIC-1: Out of memory`

Any other ideas?

**Comment by barnhill** [comment]
eww looks like a memory leak in the label generation

**Comment by tschennie** [comment]
Thats means? Is the string to encode too long? What can I change?

**Comment by barnhill** [comment]
how long is the string to encode?

**Comment by barnhill** [comment]
How long is the string you are encoding?

**Comment by barnhill** [comment]
Ok just making sure ... that should definitely be fine unless that image is MASSIVE.  I should probably put some hard limits on the sizes that are able to be specified as the size.

**Comment by tschennie** [comment]
No it's just a barcode on a shipping slip.. But sometimes it must be printed many shipping slips (with barcode on it) in series. Could that also be the problem?

**Comment by barnhill** [comment]
That would only be the problem if you didnt release the bitmap.  Then it would run out of memory like this.

**Comment by barnhill** [comment]
int Lenght = 1000 is gonna be a BIG ole image and would be long in an out of control loop to cause an out of memory error.  A side note too ... only generate the image to be the size you need to put directly on there as resizing images can cause blurring along the edges of black and white bars and this can cause scanners not to read them.  This is inherent to image resizing though and Im sure you knew that already but figured Id note it.

**Comment by barnhill** [comment]
@tschennie  one more thought on this is you could always use a 'using' statement in C# for that memory stream so its dispose is called as soon as you exit the using block.

**Comment by barnhill** [comment]
@tschennie @binki any new updates on this?

**Comment by barnhill** [comment]
@tschennie @binki I have identified a major leak.  The library incorrectly implements the IDisposable interface.  I took @tschennie code and created a for loop generating 10k barcodes.  It would spike in memory and never release it.  I have implemented and released resources in a few places that I found issues. Now instead of those causing memory to spike to over 1 gig in less than 5 seconds, I ran all 10k generations in 41mb of Ram according to JetBrains dotMemory profiler.  Im going to fix this ASAP.  OMG Im sorry it took this long to get to this.
